packages/edit-site/src/components/block-editor/block-editor-provider/use-page-content-blocks.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+import { createBlock } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { PAGE_CONTENT_BLOCK_TYPES } from '../../../utils/constants';
+
+/**
+ * Helper method to iterate through all blocks, recursing into allowed inner blocks.
+ * Returns a flattened object of transformed blocks.
+ *
+ * @param {Array}    blocks    Blocks to flatten.
+ * @param {Function} transform Transforming function to be applied to each block. If transform returns `undefined`, the block is skipped.
+ *
+ * @return {Array} Flattened object.
+ */
+function flattenBlocks( blocks, transform ) {
+	const result = [];
+	for ( let i = 0; i < blocks.length; i++ ) {
+		// Since the Query Block could contain PAGE_CONTENT_BLOCK_TYPES block types,
+		// we skip it because we only want to render stand-alone page content blocks in the block list.
+		if ( [ 'core/query' ].includes( blocks[ i ].name ) ) {
+			continue;
+		}
+		const transformedBlock = transform( blocks[ i ] );
+		if ( transformedBlock ) {
+			result.push( transformedBlock );
+		}
+		result.push( ...flattenBlocks( blocks[ i ].innerBlocks, transform ) );
+	}
+
+	return result;
+}
+
+/**
+ * Returns a memoized array of blocks that contain only page content blocks,
+ * surrounded by a group block to mimic the post editor.
+ *
+ * @param {Array}   blocks               Block list.
+ * @param {boolean} isPageContentFocused Whether the page content has focus. If `true` return page content blocks. Default `false`.
+ *
+ * @return {Array} Page content blocks.
+ */
+export default function usePageContentBlocks(
+	blocks,
+	isPageContentFocused = false
+) {
+	return useMemo( () => {
+		if ( ! isPageContentFocused || ! blocks || ! blocks.length ) {
+			return [];
+		}
+		return [
+			createBlock(
+				'core/group',
+				{
+					layout: { type: 'constrained' },
+					style: {
+						spacing: {
+							margin: {
+								top: '4em', // Mimics the post editor.
+							},
+						},
+					},
+				},
+				flattenBlocks( blocks, ( block ) => {
```

Comment:
I find it weird that we're trying to guess the blocks from the content of the template rather than just fixing "core/post-title" + "core/post-content" (like the post editor).

Reply:
Could you help me understand that statement? Is there something more reliable that could be done here?

I imagine it was approached this way to cover the (possibly unlikely) scenario that there are 2 x site titles or whatever in a template. 

Plus there is a bit of guessing required since we don't know how deep or where these blocks are located in a template in advance.

Reply:
> rather than just fixing "core/post-title" + "core/post-content" (like the post editor).

By grabbing the real blocks we ensure that the order of the content blocks matches the template so that it more closely resembles the site. I suppose the main value at the moment is that it determines whether or not a featured image is shown, and the order of the title, featured image, and content blocks. While reviewing, I thought it added a nice level of polish.

Reply:
> By grabbing the real blocks we ensure that the order of the content blocks matches the template so that it more closely resembles the site

Is that the role of this mode though? I mean, it felt like this mode is more of a "post editor" mode to me where we don't really care about the template, we're just filling the content and title.

Reply:
> I mean, it felt like this mode is more of a "post editor" mode to me where we don't really care about the template, we're just filling the content and title.

We could do it this way. 

I'd probably want to confer on the UX with some design folks.

The thinking was that, because we're in the site editor, it would be a more intuitive, seamless experience to have the page content reflect the page template.

So for instance, if I edit the Pages template in 2024 to have the post title appear underneath the post content, (just because I can) then when I edit that page in the site editor, the page corresponds with the new template structure. I can see my changes. It might be confusing to have the page content change just seconds after I've edited the template, whereas in the post editor doesn't have any of the extra context of templates etc.

That's the argument for. 

Against,  it'd be easier to refactor the code to just make it match the post editor behavior, e.g., just show the required inner blocks...

```js
				[
					createBlock( 'core/post-title' ),
					createBlock( 'core/post-content' ),
				]
```

Another angle is to remove the template preview mode altogether, I'd be for this. I understand there's a "great consolidation" going on, where the site editor will be jack of all trades, but I don't know how much value hiding the template in this way adds. Maybe to some folks.

Reply:
> I don't know how much value hiding the template in this way adds. Maybe to some folks.

I'm not sure how much value it has in its current form, I but I like the principle behind the idea of a more focused writing-mode like experience akin to the post editor, that folks can use from within the site editor. However it's exposed, I think it's important that folks can still set a featured image while editing their post content, whether or not that's in the sidebar or on the editor canvas. One of the neat things about having it in the canvas is that it gets exposed in the list view for free, so it's easier to find / interact with.